File: solutions/go/crypto-square/1/crypto_square.go
package cryptosquare

import (
	"slices"
	"strings"
	"unicode"
)

func transform(f func([]rune) string, as [][]rune) []string {
	bs := []string{}
	for _, a := range as {
		bs = append(bs, f(a))
	}
	return bs
}

func filterMap(f func(rune) rune, as string) []rune {
	bs := []rune{}
	for _, a := range as {
		b := f(a)
		if b != 0 {
			bs = append(bs, b)
		}
	}
	return bs
}

func transpose(rows [][]rune) [][]rune {
	width := len(rows[0])
	r := [][]rune{}
	for i := range width {
		s := []rune{}
		for _, row := range rows {
			s = append(s, row[i])
		}
		r = append(r, s)
	}
	return r
}

func runesToString(rs []rune) string { return string(rs) }

func Encode(plainText string) string {
	message := filterMap(func(r rune) rune {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			return unicode.ToLower(r)
		}
		return 0
	}, plainText)

	c := 0
	r := 0
	for c*r < len(message) {
		if r < c {
			r++
		} else {
			c++
		}
	}
	if c == 0 {
		return ""
	}

	plainSquare := slices.Collect(slices.Chunk(message, c))
	lastLine := plainSquare[len(plainSquare)-1]
	if len(lastLine) < c {
		plainSquare[len(plainSquare)-1] = append(
			lastLine,
			slices.Repeat([]rune{' '}, c-len(lastLine))...,
		)
	}

	cipherSquare := transpose(plainSquare)
	cipherText := strings.Join(transform(runesToString, cipherSquare), " ")
	return cipherText
}

File: solutions/haskell/crypto-square/1/src/CryptoSquare.hs
{-# LANGUAGE LambdaCase #-}

module CryptoSquare (encode) where

import Data.Char (isDigit, isLetter, toLower)
import Data.List (transpose)

encode :: String -> String
encode plainText
    | null message = ""
    | otherwise = unwords cipherSquare
  where
    message = [toLower c | c <- plainText, isLetter c || isDigit c]
    messageLength = length message
    cols = head [c | r <- [0 ..], c <- [r, r + 1], c * r >= messageLength]
    plainSquare = chunksOf cols message
    lastLine = last plainSquare
    plainSquarePadded =
        init plainSquare ++ [lastLine ++ replicate (cols - length lastLine) ' ']
    cipherSquare = transpose plainSquarePadded

chunksOf :: Int -> [a] -> [[a]]
chunksOf n = \case
    [] -> []
    xs -> t : chunksOf n d where (t, d) = splitAt n xs
